src/Authorization.AspNetCore/AuthorizationValidationRule.cs

```diff
@@ -68,108 +71,120 @@ public Task<INodeVisitor> ValidateAsync(ValidationContext context)
-            if (type == null || !type.RequiresAuthorization())
-            {
-                return;
-            }
+            var policyNames = type?.GetPolicies();
 
-            var policyNames = type.GetPolicies();
-            if (policyNames.Count == 0)
+            if (policyNames?.Count == 1)
             {
-                return;
+                // small optimization for the single policy - no 'new List<>()', no 'await Task.WhenAll()'
+                var authorizationResult = await _authorizationService.AuthorizeAsync(_httpContextAccessor.HttpContext.User, policyNames[0]);
+                AddValidationError(node, context, operationType, authorizationResult);
             }
-
-            var tasks = new List<Task<AuthorizationResult>>(policyNames.Count);
-            foreach (string policyName in policyNames)
+            else if (policyNames?.Count > 1)
             {
-                var task = _authorizationService.AuthorizeAsync(_httpContextAccessor.HttpContext.User, policyName);
-                tasks.Add(task);
+                var tasks = new List<Task<AuthorizationResult>>(policyNames.Count);
+                foreach (string policyName in policyNames)
+                {
+                    var task = _authorizationService.AuthorizeAsync(_httpContextAccessor.HttpContext.User, policyName);
+                    tasks.Add(task);
+                }
+
+                var authorizationResults = await Task.WhenAll(tasks);
+
+                foreach (var result in authorizationResults)
+                {
+                    AddValidationError(node, context, operationType, result);
+                }
             }
-            await Task.WhenAll(tasks);
 
-            foreach (var task in tasks)
+            static void AddValidationError(INode node, ValidationContext context, OperationType? operationType, AuthorizationResult result)
             {
-                var result = task.Result;
                 if (!result.Succeeded)
                 {
-                    var stringBuilder = new StringBuilder("You are not authorized to run this ");
-                    stringBuilder.Append(operationType.ToString().ToLower());
-                    stringBuilder.AppendLine(".");
+                    var error = new StringBuilder("You are not authorized to run this ")
+                                .Append(GetOperationType(operationType))
+                                .Append(".");
 
                     foreach (var failure in result.Failure.FailedRequirements)
                     {
-                        AppendFailureLine(stringBuilder, failure);
+                        AppendFailureLine(error, failure);
                     }
 
-                    context.ReportError(new ValidationError(context.OriginalQuery, "6.1.1", stringBuilder.ToString(), node)
+                    context.ReportError(new ValidationError(context.OriginalQuery, "6.1.1", error.ToString(), node == null ? Array.Empty<INode>() : new INode[] { node })
```

Comment:
I have to resort to such a trick, otherwise `null` is wrapped as the single element of the array, followed by NRE when iterating over the array in `ValidationError` ctor.
![изображение](https://user-images.githubusercontent.com/21261007/101280299-737dfe80-37d9-11eb-9ca8-a3099408c2d6.png)